tests/test_save_load.py

```diff
@@ -114,6 +114,94 @@ def tiny_save_load(tiny_args, fsdp=False):
         shutil.rmtree(override_params["checkpoint_path"], ignore_errors=True)
 
 
+def tiny_save_load_different_seed(tiny_args, fsdp=False):
+    """
+    This test checks that the model will not resume if the seed is different.
+    """
+    scaler = None
+    epoch = 0
+    evaluation_metrics = None
+    global_step = 0
+    done_training = False
+    download_dl_test_data()
+    override_params = dict(
+        seq_len=tiny_args.sequence_length,
+        vocab_size=tiny_args.vocab_size,
+        train_num_samples=tiny_args.train_num_samples,
+        global_batch_size=tiny_args.global_batch_size,
+        checkpoint_path="./tests/assets/checkpoints/tiny_model/",
+        device="cpu" if not torch.cuda.is_available() else "cuda",
+        dataset_type="synthetic",
+        save_logs=True,
+        distributed=fsdp or is_using_distributed(),
+        force_distributed=fsdp or is_using_distributed(),
+        seed=0,
+    )
+
+    try:
+        args, model, data, optimizer, scheduler, loss = create_train_fixtures(
+            tiny_args.model, fsdp=fsdp, **override_params
+        )
+        model = model.to(args.device)
+        args2, model2, data2, optimizer2, scheduler2, loss2 = create_train_fixtures(
+            tiny_args.model, fsdp=fsdp, **override_params
+        )
+        model2 = model2.to(args2.device)
+        os.makedirs(args.checkpoint_path, exist_ok=True)
+
+        # print("Training tiny model")
+        train_one_epoch(
+            model,
+            data,
+            CrossEntropyLossWithZLoss(),
+            epoch=epoch,
+            step=global_step,
+            optimizer=optimizer,
+            scaler=scaler,
+            scheduler=scheduler,
+            total_steps=args.train_num_samples // args.global_batch_size,
+            args=args,
+        )
+        epoch += 1
+        threshold = 1e-6
+        # Checking that tiny models diverged after traning.
+        allclose = True
+        for (n1, p1), (n2, p2) in zip(model.named_parameters(), model2.named_parameters()):
+            allclose = allclose and torch.allclose(p1, p2, atol=threshold)
+        assert not allclose
+
+        print("Saving tiny model")
+        # Saving checkpoints.
+        save_checkpoint(
+            args,
+            model,
+            optimizer,
+            scaler,
+            epoch,
+            evaluation_metrics,
+            step=global_step,
+            is_final_checkpoint=done_training,
+            next_shard_per_source=None,
+            samples_seen=None,
+            shard_shuffle_seed=args.seed,
+        )
+
+        # Loading saved tiny model - this should fail because the seed also changes
```

Comment:
is there a way to check that the sample is correct? like if i trained 5 shards, i should start on the 6th

Reply:
The added tests ensure determinism across both multiple calls of `get_string_for_epoch` with the same seed and across resumes.